lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,722 @@
+        """Handle the async_primary relation being changed."""
+        state = self.get_state(event.relation)
+
+        if state == States.INITIALIZING:
+            # Add replica cluster primary node
+            logger.debug("Initializing replica cluster")
+            self._charm.unit.status = MaintenanceStatus("Adding replica cluster")
+            remote_data = self.get_remote_relation_data(event.relation) or {}
+
+            cluster = remote_data["cluster-name"]
+            endpoint = remote_data["endpoint"]
+            unit_label = remote_data["node-label"]
+
+            if cluster == self.cluster_name:
+                import uuid
+
+                logger.warning(
+                    "Cluster name is the same as the primary cluster. Appending generetade value"
+                )
+                cluster = f"{cluster}{uuid.uuid4().hex[:4]}"
+
+            logger.debug("Looking for a donor node")
+            _, ro, _ = self._charm._mysql.get_cluster_endpoints(get_ips=False)
+
+            if not ro:
+                logger.debug(f"Adding replica {cluster=} with {endpoint=}. Primary is the donor")
+                self._charm._mysql.create_replica_cluster(
+                    endpoint, cluster, instance_label=unit_label
+                )
+            else:
+                donor = ro.split(",")[0]
+                logger.debug(f"Adding replica {cluster=} with {endpoint=} using {donor=}")
+                self._charm._mysql.create_replica_cluster(
+                    endpoint, cluster, instance_label=unit_label, donor=donor
+                )
+
+            event.relation.data[self.model.app]["replica-state"] = "initialized"
+            logger.debug("Replica cluster created")
+            self._charm._on_update_status(None)
+
+        elif state == States.RECOVERING:
+            # Recover replica cluster
+            self._charm.unit.status = MaintenanceStatus("Replica cluster in recovery")
+
+
+class MySQLAsyncReplicationReplica(MySQLAsyncReplication):
+    """MySQL async replication replica side.
+
+    Implements the setup phase of the async replication for the replica side.
+    """
+
+    def __init__(self, charm: "MySQLOperatorCharm"):
+        super().__init__(charm, REPLICA_RELATION)
+
+        if self._charm.unit.is_leader():
+            # leader/primary
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_created, self._on_replica_created
+            )
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_changed, self._on_replica_changed
+            )
+        else:
+            # non-leader/secondaries
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_created,
+                self._on_replica_secondary_created,
+            )
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_changed,
+                self._on_replica_secondary_changed,
```

Comment:
this is a bit confusing: should secondary be renamed to `non_leader` here if it is in juju sense?

Reply:
Agreed. Changed